integrations/notification_helper_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package integrations
+
+import (
+	"encoding/json"
+	"reflect"
+	"sync"
+	"testing"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification"
+	"code.gitea.io/gitea/modules/notification/base"
+	"code.gitea.io/gitea/modules/queue"
+)
+
+var notifierListener *NotifierListener
+
+var once = sync.Once{}
+
+type NotifierListener struct {
+	lock      sync.RWMutex
+	callbacks map[string][]*func(string, [][]byte)
+	notifier  base.Notifier
+}
+
+func NotifierListenerInit() {
+	once.Do(func() {
+		notifierListener = &NotifierListener{
+			callbacks: map[string][]*func(string, [][]byte){},
+		}
+		notifierListener.notifier = base.NewQueueNotifierWithHandle("test-notifier", notifierListener.handle)
+		notification.RegisterNotifier(notifierListener.notifier)
+	})
+}
+
+// Register will register a callback with the provided notifier function
+func (n *NotifierListener) Register(functionName string, callback *func(string, [][]byte)) {
+	n.lock.Lock()
+	n.callbacks[functionName] = append(n.callbacks[functionName], callback)
+	n.lock.Unlock()
+}
+
+// Deregister will remove the provided callback from the provided notifier function
+func (n *NotifierListener) Deregister(functionName string, callback *func(string, [][]byte)) {
+	n.lock.Lock()
+	found := -1
+	for i, callbackPtr := range n.callbacks[functionName] {
+		if callbackPtr == callback {
+			found = i
+			break
+		}
+	}
+	if found > -1 {
+		n.callbacks[functionName] = append(n.callbacks[functionName][0:found], n.callbacks[functionName][found+1:]...)
```

Comment:
Since you will be breaking from the loop, I think it's safe to omit the `found` variable and move this inside `if callbackPtr == callback`.

Reply:
Done